OneCommunityGlobal/HGNRest
# Ilya - Tools table

<img width="771" alt="tool-list-task" src="https://github.com/OneCommunityGlobal/HGNRest/assets/105448137/1c6dc9fc-7081-419c-aeb6-558fcc11ad42">

# Description
Adds functionality for the user to navigate to [/bmdashboard/tools](http://localhost:3000/bmdashboard/tools), and view a list of all tool items currently in the system. The table can be sorted by tool name or by project and the results can be filtered by tool name. Update History and Purchase History can be viewed for each item in a modal.

## Related PRS (if any):
This backend PR is related to the [#2193](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2193) frontend PR.
To test this backend PR you need to checkout the #2193 frontend PR, branch **ilya_tools_table_fe**

## Main changes explained:
- largeItemBaseSchema changed to have a "condition" field
- buildingTool schema changed to have specific values for the purchaseStatus field
- fetchAllTools function added to bmToolController to query the database and return a full list of tool items and populate certain fields
- A new route added to call fetchAllTools when a GET request is made to /api/tools

## How to test:
1. Check out the current branch. Make sure you are using Node version 14.17.6. Run **npm install** and **npm run build && npm start** to run the backend locally
2. Check out **ilya_tools_table_fe** branch on the front end. Run commands **npm install** and **npm run start:local** to launch the front end application
3. On the front end log in as any user
4. Go to [/bmdashboard](http://localhost:3000/bmdashboard) and log in again
5. Navigate to [/bmdashboard/tools](http://localhost:3000/bmdashboard/tools). A list of all tools currently in the system should display.
6. Test the filtering by Project and Tool function (the dropdowns above the table)
7. Test sorting by project and by name (the arrows next to "Project" and "Name" in the table header
8. Pressing the View button under Updates should open an Updates Record modal (not all items have information to display)
9. Pressing the View button under Purchases should open an Purchase Record modal (not all items have information to display)
10. Pressing the pencil should open an Update Tool Form modal. The content of this modal is outside of the scope of this task

## Screenshots or video

https://github.com/OneCommunityGlobal/HGNRest/assets/105448137/891fe190-e5c4-492d-86c0-c7362769957a
